ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/ScalablePushRegistry.java

```diff
@@ -0,0 +1,177 @@
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.physical.scalablepush;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.GenericKey;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.streams.materialization.Row;
+import io.confluent.ksql.execution.streams.materialization.TableRow;
+import io.confluent.ksql.execution.streams.materialization.WindowedRow;
+import io.confluent.ksql.physical.scalablepush.locator.AllHostsLocator;
+import io.confluent.ksql.physical.scalablepush.locator.PushLocator;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import io.vertx.core.impl.ConcurrentHashSet;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This registry is kept with every persistent query, peeking at the stream which is the output
+ * of the topology. These rows are then fed to any registered ProcessingQueues where they are
+ * eventually passed on to scalable push queries.
+ */
+public class ScalablePushRegistry implements ProcessorSupplier<Object, GenericRow> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ScalablePushRegistry.class);
+
+  private final PushLocator pushLocator;
+  private final LogicalSchema logicalSchema;
+  private final boolean windowed;
+  private final Set<ProcessingQueue> processingQueues = new ConcurrentHashSet<>();
+  private volatile boolean closed = false;
+
+  public ScalablePushRegistry(
+      final PushLocator pushLocator,
+      final LogicalSchema logicalSchema,
+      final boolean windowed
+  ) {
+    this.pushLocator = pushLocator;
+    this.logicalSchema = logicalSchema;
+    this.windowed = windowed;
+  }
+
+  public void close() {
+    for (ProcessingQueue queue : processingQueues) {
+      queue.close();
+    }
+    processingQueues.clear();
+    closed = true;
+  }
+
+  public void register(final ProcessingQueue processingQueue) {
+    if (closed) {
```

Comment:
I think there's a race condition here - if someone calls `close` and another calls `register` then close can be in the middle of iterating when we add another one to the set. I can't tell whether this is meant to be concurrent code or not, but the fact the `closed` is volatile makes me think it is. It seems to me like we'd be safer off just synchronizing everything

Reply:
The code that calls this effectively shouldn't ever register anything after calling close, so it's a bit funny to reason about what it means to be consistent, but I agree that it's confusing this way.  I added synchronized to protect the use of `closed` and changing the size of `processingQueues`.

I still use a ConcurrentHashMap for `processingQueues` so that I don't need to get a lock when processing rows.  It gives a weakly consistent view of the map while iterating, which should be performant.